src/webgl/framebuffer.js

```diff
@@ -1,67 +1,48 @@
-import {GL, glGet, glTypeToArray, glTypeFromArray} from './webgl';
-import {assertWebGLContext, assertWebGL2Context} from './webgl-checks';
+import GL from './api';
+import {assertWebGLContext, assertWebGL2Context} from './context';
+import Resource from './resource';
 import Texture2D from './texture-2d';
 import Renderbuffer from './renderbuffer';
+import {getTypedArrayFromGLType, getGLTypeFromTypedArray} from '../utils/typed-array-utils';
+import {log} from '../utils';
 import assert from 'assert';
-import {uid, log} from '../utils';
 
-// Returns number of components in a specific WebGL format
-function glFormatToComponents(format) {
-  switch (format) {
-  case GL.ALPHA: return 1;
-  case GL.RGB: return 3;
-  case GL.RGBA: return 4;
-  default: throw new Error('Unknown format');
-  }
-}
+const ATTACHMENT_PARAMETERS = [
+  GL.FRAMEBUFFER_ATTACHMENT_OBJECT_NAME, // WebGLRenderbuffer or WebGLTexture
+  GL.FRAMEBUFFER_ATTACHMENT_OBJECT_TYPE, // GL.RENDERBUFFER, GL.TEXTURE, GL.NONE
+  GL.FRAMEBUFFER_ATTACHMENT_TEXTURE_CUBE_MAP_FACE, // GL.TEXTURE_CUBE_MAP_POSITIVE_X, etc.
+  GL.FRAMEBUFFER_ATTACHMENT_TEXTURE_LEVEL, // GLint
+  // EXT_sRGB or WebGL2
+  GL.FRAMEBUFFER_ATTACHMENT_COLOR_ENCODING, // GL.LINEAR, GL.SRBG
+  // WebGL2
+  GL.FRAMEBUFFER_ATTACHMENT_TEXTURE_LAYER, // GLint
+  GL.FRAMEBUFFER_ATTACHMENT_RED_SIZE, // GLint
+  GL.FRAMEBUFFER_ATTACHMENT_GREEN_SIZE, // GLint
+  GL.FRAMEBUFFER_ATTACHMENT_BLUE_SIZE, // GLint
+  GL.FRAMEBUFFER_ATTACHMENT_ALPHA_SIZE, // GLint
+  GL.FRAMEBUFFER_ATTACHMENT_DEPTH_SIZE, // GLint
+  GL.FRAMEBUFFER_ATTACHMENT_STENCIL_SIZE, // GLint
+  GL.FRAMEBUFFER_ATTACHMENT_COMPONENT_TYPE
+  // GL.FLOAT, GL.INT, GL.UNSIGNED_INT, GL.SIGNED_NORMALIZED, OR GL.UNSIGNED_NORMALIZED.
+];
+
+export default class Framebuffer extends Resource {
 
-export default class Framebuffer {
-  static makeFrom(gl, object = {}) {
-    return object instanceof Framebuffer ? object :
-      // Use .handle (e.g from stack.gl's gl-buffer), else use buffer directly
-      new Framebuffer(gl, {handle: object.handle || object});
-  }
-
-  /* eslint-disable max-statements */
   constructor(gl, opts = {}) {
-    assertWebGLContext(gl);
-
-    const {id} = opts;
-
-    const handle = gl.createFramebuffer();
-    if (!handle) {
-      throw new Error('Failed to create WebGL Framebuffer');
-    }
-
-    this.gl = gl;
-    this.id = uid(id);
-    this.handle = handle;
-    this.colorBuffer = null;
-    this.depthBuffer = null;
-    this.stencilBuffer = null;
-    this.texture = null;
-    this.userData = {};
-    this.width = 0;
-    this.height = 0;
+    super(gl, opts);
+    this.initialize(opts);
     Object.seal(this);
-
     this.resize(opts);
```

Comment:
I don't really get it... why do we call resize (which is essentially a this.initialize() call,  right after another this.initilize() call?